### ISSUE TITLE
Compiling URDF and SRDF parsers

### ISSUE THREAD

**Issue by jcarpent** [stack-of-tasks/pinocchio#1078]
Current implementations of URDF and SRDF parsers are header-only, meaning that the private flags related to the various versions of URDFDOM are transferred for correct compilation. In addition, for each new executable, we have to compile the signatures again and again, while by default, URDF and SRDF assumes double scalar type.

I would like to pass these parsers to the existing but empty Pinocchio. This lib will only contain the parsers signatures, nothing more. @jmirabel @wxmerkt @cmastalli @nim65s @nmansard @aelkhour 
What is your opinion on this suggestion?

**Comment by jmirabel** [comment]
I like the idea but I fear it is gonna be hard to achieve. In HPP, we are using a different JointModelCollection. I think this would be an issue, right ?

**Comment by jcarpent** [comment]
No, because the JointCollection is also in the signature of default Model in Pinocchio. 
But, this means that we still need to export the compilations flags.

**Comment by jcarpent** [comment]
Another solution would be to proceed to a cast.

**Comment by jmirabel** [comment]
> No, because the JointCollection is also in the signature of default Model in Pinocchio.

That's the point. We don't use the default Model signature.

**Comment by jmirabel** [comment]
We could also use a visitor design pattern.

**Comment by jcarpent** [comment]
What do you mean?

**Comment by jmirabel** [comment]
Add a class UrdfVisitor that has a prototype such as:
```c++
class UrdfVisitor {
virtual void addFreeFlyer(/* relevant args, excluding the model */) = 0;
virtual void addRevolute(/* relevant args, excluding the model */) = 0;
// and one more for each type of joint

virtual void addLink(/* relevant args, excluding the model */) = 0;
// and any other method that should be exposed.
};
```

Then a child class
```
template <typename Model> ChildClass
{
Model& model;
// implementation of each of the parent virtual methods.
}
```

Then, the URDF parser becomes independant of class Model.

**Comment by jcarpent** [comment]
My initial goal would be to not have to export URDF related flags when compiling Pinocchio. I don't see how it will happen in your case?

**Comment by jmirabel** [comment]
In the library, you will have a code that depends on URDF related flags like this:
```c++
void parse(std::string filename, UrdfVisitor* visitor)
{
UrdfTree urdf (filename);
...
// Add a joint
visitor->addRevolute (position_in_parent, axis, limits...);
}
```
Then, the UrdfVisitor::addRevolute function will not depend on anything related to urdfdom. You should not need to include any urdfdom header.

**Comment by jcarpent** [comment]
OK. Thanks for the clarification. As you seems much more aware on the topic, would you have time to implement it?

**Comment by jmirabel** [comment]
I can put this onto my task list but it won't have a high priority. There is no hurry for this, is there ?

**Comment by jcarpent** [comment]
No, of course.

**Comment by jcarpent** [comment]
Thanks a lot @jmirabel for handling this feature.

**Comment by jcarpent** [comment]
There is still the SRDF parser to do.

**Comment by jmirabel** [comment]
You are right. I though the intend was to get rid of *urdfdom* stuff in the installed includes. The SRDF parser only relies on boost, and not on *urdfdom*.

**Comment by jcarpent** [comment]
@nim65s or @proyan Could you handle this (SRDF)?

**Comment by proyan** [comment]
This issue is in liaison with https://github.com/stack-of-tasks/pinocchio/issues/1498, and normally this should be resolved when 1498 is handled

**Comment by jcarpent** [comment]
@proyan Could you handle this issue in pinocchio-3x?
